### PR TITLE
build: make dynamic linking opt-in instead of default

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- It seems dynamic linking still needs configuration (especially OpenSSL), so it might be better to static link by default.
- Still dynamic linking in CI debug build. It's 30s faster

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: docs/developer-guide.md [hunk 1]
```diff
@@ -73,19 +73,18 @@ RiseDev is the development mode of RisingWave. To develop RisingWave, you need t
...
-brew install postgresql cmake protobuf tmux cyrus-sasl llvm zstd
+brew install postgresql cmake protobuf tmux cyrus-sasl llvm
 curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
 ```
 
 To install the dependencies on Debian-based Linux systems, run:
 
 ```shell
-sudo apt install make build-essential cmake protobuf-compiler curl openssl libssl-dev libsasl2-dev libcurl4-openssl-dev pkg-config postgresql-client tmux lld libzstd-dev
+sudo apt install make build-essential cmake protobuf-compiler curl openssl libssl-dev libsasl2-dev libcurl4-openssl-dev pkg-config postgresql-client tmux lld
```
Comment (xxchan) [inline docs/developer-guide.md:87]: I think more can be removed, but I'm not brave enough to do it. 😅

File: src/risedevtool/config/src/main.rs
Comment (xxchan) [inline src/risedevtool/config/src/main.rs:1]: Removing so that it compiles fast & won't fail.

File: Makefile.toml [hunk 2]
```diff
@@ -22,11 +22,16 @@ env_scripts = [
...
 
+# only duckscript can modify env variables in cargo-make
+
 set_env ENABLE_TELEMETRY "false"
 
 is_sanitizer_enabled = get_env ENABLE_SANITIZER
 is_all_in_one_enabled = get_env ENABLE_ALL_IN_ONE
 is_hdfs_backend = get_env ENABLE_HDFS
+is_release = get_env ENABLE_RELEASE_PROFILE
+is_release = set ${is_release} or false
```
Comment (xxchan) [inline Makefile.toml:33]: workaround https://github.com/sagiegurari/duckscript/issues/326

File: Makefile.toml [hunk 3]
```diff
@@ -329,7 +340,7 @@ echo + RUSTFLAGS="${RUSTFLAGS:-<not set>}"
 set -xe
 cargo build -p ${RISEDEV_CARGO_BUILD_CRATE} -p risedev ${BUILD_HDFS_BACKEND_CMD}\
             --profile "${RISINGWAVE_BUILD_PROFILE}" \
-            --features "${RISINGWAVE_FEATURES}" \
+            ${RISINGWAVE_FEATURE_FLAGS} \
```
Comment (xxchan) [inline Makefile.toml:343]: I considered make `rw-static-link` default feature, and add `--no-default-features` for dynamic linking. But not sure.

File: src/utils/workspace-config/Cargo.toml [hunk 4]
```diff
@@ -9,14 +9,15 @@ repository = { workspace = true }
...
+# so they are two features :)
+rw-static-link = ["openssl-sys", "sasl2-sys"]
 rw-dynamic-link = ["zstd-sys"]
 
 [dependencies]
 log = { version = "0.4", features = ["release_max_level_debug"] }
-openssl = { version = "0.10", optional = true, features = ["vendored"] }
-rdkafka = { workspace = true, optional = true, features = ["cmake-build", "ssl-vendored", "gssapi-vendored"] }
+openssl-sys = { version = "0.9", optional = true, features = ["vendored"] }
+sasl2-sys = { version = "0.1", optional = true, features = ["gssapi-vendored"] }
```
Comment (xxchan) [inline src/utils/workspace-config/Cargo.toml:20]: I think if *any* crate enabled vendored ssl, the eventual result is enabling `openssl-sys/vendored`. So I just use its own feature here to be more transparent.

File: src/utils/workspace-config/Cargo.toml [hunk 5]
```diff
@@ -9,14 +9,15 @@ repository = { workspace = true }
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [features]
-# Use a feature flag to control enable or not, otherwise `cargo test` will compile all dependencies.
```
Comment (xxchan) [inline src/utils/workspace-config/Cargo.toml:12]: I don’t understand the comment, so removed 😅

File: Makefile.toml [hunk 6]
```diff
@@ -49,6 +54,12 @@ else
   set_env BUILD_HDFS_BACKEND_CMD ""
 end
 
+if ${is_not_release} and ${is_dynamic_linking}
```
Comment (xxchan) [inline Makefile.toml:57]: Note: `if not ${is_release} and ${is_dynamic_linking}` doesn't work. https://github.com/sagiegurari/duckscript/issues/326

File: docs/developer-guide.md (hunk 1 above)
Comment (BugenZhao) [reply docs/developer-guide.md:87]: I guess this can be aligned with `ci/Dockerfile`. 👀

File: src/utils/workspace-config/Cargo.toml (hunk 4 above)
Comment (BugenZhao) [reply src/utils/workspace-config/Cargo.toml:20]: LGTM!

File: Makefile.toml (hunk 3 above)
Comment (BugenZhao) [reply Makefile.toml:343]: Both LGTM as long as we handle and hide all this stuff in RiseDev. 🤣

**Comment by xxchan** [comment]
@BugenZhao request approval

**Review by BugenZhao — APPROVED** [review]
LGTM

**Pull request merged**